src/Hosting/WindowsService/src/ServiceBaseLifetime.cs

```diff
@@ -1,41 +1,42 @@
-﻿#if NET461
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
 using System;
 using System.ServiceProcess;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.Extensions.DependencyInjection;
-using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
 
-namespace GenericHostSample
+namespace Microsoft.Extensions.Hosting.WindowsService
 {
-    public static class ServiceBaseLifetimeHostExtensions
-    {
-        public static IHostBuilder UseServiceBaseLifetime(this IHostBuilder hostBuilder)
-        {
-            return hostBuilder.ConfigureServices((hostContext, services) => services.AddSingleton<IHostLifetime, ServiceBaseLifetime>());
-        }
-
-        public static Task RunAsServiceAsync(this IHostBuilder hostBuilder, CancellationToken cancellationToken = default)
-        {
-            return hostBuilder.UseServiceBaseLifetime().Build().RunAsync(cancellationToken);
-        }
-    }
-
     public class ServiceBaseLifetime : ServiceBase, IHostLifetime
     {
         private TaskCompletionSource<object> _delayStart = new TaskCompletionSource<object>();
 
-        public ServiceBaseLifetime(IApplicationLifetime applicationLifetime)
+        public ServiceBaseLifetime(IHostingEnvironment environment, IApplicationLifetime applicationLifetime, ILoggerFactory loggerFactory)
         {
+            Environment = environment ?? throw new ArgumentNullException(nameof(environment));
             ApplicationLifetime = applicationLifetime ?? throw new ArgumentNullException(nameof(applicationLifetime));
+            Logger = loggerFactory.CreateLogger("Microsoft.Hosting.Lifetime");
         }
 
+        public IHostingEnvironment Environment { get; }
         private IApplicationLifetime ApplicationLifetime { get; }
+        private ILogger Logger { get; }
 
         public Task WaitForStartAsync(CancellationToken cancellationToken)
         {
             cancellationToken.Register(() => _delayStart.TrySetCanceled());
-            ApplicationLifetime.ApplicationStopping.Register(Stop);
+            ApplicationLifetime.ApplicationStarted.Register(() =>
+            {
+                Logger.LogInformation("Application started. Hosting environment: {envName}; Content root path: {contentRoot}",
+                    Environment.EnvironmentName, Environment.ContentRootPath);
+            });
+            ApplicationLifetime.ApplicationStopping.Register(() =>
```

Comment:
Remind me why we need this and why handling Stop isn't enough?

Reply:
So you can initiate shutdown from within the app.

Reply:
What case do you need this?

https://github.com/aspnet/Extensions/blob/b560068287b890594046de034db2c28c278f19e9/src/Hosting/Abstractions/src/HostingAbstractionsHostExtensions.cs#L97-L106

Is it only useful when you call Start on the host?

Reply:
That one covers CTL+C, service stop, applicationLifetime.StopApplication(), etc..

Reply:
I'm confused. What do you mean? What case does this cover that that one does not? The reason I'm asking is because I wouldn't expect every lifetime to do this (it's not part of the contract which means there's s design flaw somewhere)

Reply:
Stopping can be called from a lot of places, some of which are circular. The various lifetimes also have things they want to do when the app is stopping, such as log, so they listen to the same event as they may or may not have triggered it.

Here are a few flows:
- Console lifetime, CTL+C or ProcessExit -> ApplicationLifetime.StopApplication() -> unblock WaitForShutdownAsync -> host.StopAsync
- ServiceBaseLifetime service Stop -> ApplicationLifetime.StopApplication() -> unblock WaitForShutdownAsync -> host.StopAsync
- Application code -> ApplicationLifetime.StopApplication() -> unblock WaitForShutdownAsync -> host.StopAsync -> ServiceBaseLifetime.StopAsync -> ServiceBase.Stop
 - Full control Program.Main - Doesn't use WaitForShutdown, calls Host.StopAsync directly -> ServiceBaseLifetime.StopAsync -> ServiceBase.Stop

Ok, so ServiceBaseLifetime might not need ApplicationStopping to call Stop, but it still wants it for logging just like ConsoleLifetime because ApplicationStopping fires at the start of the shutdown process.